### PR TITLE
Core Data: Optimize storage usage for product attribute stores

### PR DESCRIPTION
**Pull request by itsmeichigo**
Part of #14091 

## Description

This PR continue the Core Data optimization project to improve the storage usage in `ProductAttributeStore` and `ProductAttributeTermStore`. Changes include:
- Replace the deprecated access to `writerDerivedStorage` with the new method `performAndSave` in `upsertStoredProductAttributesInBackground` and `upsertStoredProductAttributeTermsInBackground`.
- Updated `upsertStoredProductAttributesInBackground` to optionally delete stale items.
- Moved the deletion of attributes and attribute terms to the background.
- Optimize fetch requests in upsertion methods to fetch a group of items instead of separate fetch requests in for loops.

## Steps to reproduce

- Log in to a test store.
- Navigate to the Products tab and select to manually create a new variable product.
- On the product form, select Add variations > Add attributes.
- Enter a new attribute name, then enter different options for that attribute. E.g. Flavor > strawberry / orange / mango.
- Select Next > Generate variations. Confirm that the generation succeeds. 
- Add more details as needed and save the product.
- Select the product again > select Variation Attributes, and confirm that the correct attributes are listed.
- Select any of the attributes, and confirm that the correct options are listed.

## Testing information

Tested on simulator iPhone 16 Pro iOS 18.0 and confirmed for a variable product:
- Creating attributes works as expected.
- Creating attribute terms works as expected.
- Fetching details for attributes and attribute terms for variable products work as expected.

## Screenshots

N/A
---
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

TODO: @itsmeichigo to update the release notes before merging.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

### CODE REVIEW

File: Yosemite/Yosemite/Stores/ProductAttributeTermStore.swift [hunk 1]
```diff
@@ -61,17 +57,18 @@ private extension ProductAttributeTermStore {
...
-                self.deleteStaleTerms(siteID: siteID, attributeID: attributeID, activeTerms: synchronizedTerms + terms)
-                onCompletion(nil)
+                deleteStaleTerms(siteID: siteID, attributeID: attributeID, activeTerms: synchronizedTerms + terms) {
+                    onCompletion(nil)
+                }
 
             // If there could be more(non-empty) terms, request the next page recursively.
             case let .success(terms):
-                self.synchronizeAllProductAttributeTerms(siteID: siteID,
+                synchronizeAllProductAttributeTerms(siteID: siteID,
```
Comment (selanthiraiyan) [inline Yosemite/Yosemite/Stores/ProductAttributeTermStore.swift:71]: Super nit: Indentation fix needed.

**Review by selanthiraiyan — APPROVED** [review]
The code LGTM and product attribute generation work as described. :shipit:

File: Yosemite/Yosemite/Stores/ProductAttributeTermStore.swift (hunk 1 above)
Comment (itsmeichigo) [reply Yosemite/Yosemite/Stores/ProductAttributeTermStore.swift:71]: Done in 7920636dcc371e26917c35e6df9010e84f44d062.

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/229">20.9</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Pull request merged**